### PR TITLE
Changed: Multicomponent material models now assume compositional field values are mass fractions

### PR DESCRIPTION
**Pull request by bobmyhill**
The multicomponent and multicomponent_compressible material models currently assume that the compositional fields correspond to volume fractions. This is problematic because ASPECT's governing equations assume that the compositional fields are locally conserved. In this PR, these material models now assume that the compositional fields correspond to mass fractions. Material properties were also averaged differently in the two material models, they have now been unified.

Finally, this PR changes the name of the MaterialUtilities::compute_volume_fractions function to MaterialUtilities::compute_fractions_from_compositional_fields. The old name was misleading. 

* [x] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).
* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

File: source/material_model/multicomponent.cc
Comment (jdannberg) [inline source/material_model/multicomponent.cc:186]: This model doesn't have compressibility.

File: tests/multicomponent_compressible_averaging/solution/solution-00000.0000.gnuplot
Comment (jdannberg) [inline tests/multicomponent_compressible_averaging/solution/solution-00000.0000.gnuplot:9]: Hm. Interesting. In this test, all the volume-average properties look good, but the specific heat is different from what we expect.

**Review by jdannberg — COMMENTED** [review]
Sorry for taking so long to review this. Thank you for implementing this!

First, I have a general question. In compressible models, our compositional fields are usually mass fractions (unless we add additional terms). But in an incompressible model, isn't it somewhat arbitrary if they are mass or volume fractions (because div u = 0)? Of course it still makes a difference for the computation, because the density is not constant. So since this pull request changes the results, do we all agree this should be mass fractions? @naliboff? @gassmoeller?

Then I have three major points I think we need to think about:
(1) I am a bit surprised by the test result (they don't agree with the reference solution you provide). Can you look into this? If this turns out to be complicated, I would just change the test to have a constant temperature. 
(2) Have you looked at how much the test results change? Does this make a big difference anywhere, or are these all small changes?
(3) What do we want to do about all the other material models that use this functionality, but still assume volume fractions?

File: benchmarks/newton_solver_benchmark_set/spiegelman_et_al_2016/drucker_prager_compositions.cc
Comment (bobmyhill) [reply benchmarks/newton_solver_benchmark_set/spiegelman_et_al_2016/drucker_prager_compositions.cc:264]: This is already an open issue (#3827), but I've added a comment describing the function rename.

File: doc/modules/changes/20200828b_bobmyhill
Comment (bobmyhill) [reply doc/modules/changes/20200828b_bobmyhill:13]: Done.

File: benchmarks/nonlinear_channel_flow/simple_nonlinear.cc
Comment (bobmyhill) [reply benchmarks/nonlinear_channel_flow/simple_nonlinear.cc:144]: See above.

File: include/aspect/material_model/utilities.h
Comment (bobmyhill) [reply include/aspect/material_model/utilities.h:230]: I've added the following sentence at the end of the docstring: 
> This function makes no assumptions about the units of the compositional field values; for example, they could correspond to mass or volume fractions.

File: include/aspect/material_model/equation_of_state/multicomponent_compressible.h
Comment (bobmyhill) [reply include/aspect/material_model/equation_of_state/multicomponent_compressible.h:38]: Thanks :D

File: include/aspect/material_model/equation_of_state/multicomponent_compressible.h
Comment (bobmyhill) [reply include/aspect/material_model/equation_of_state/multicomponent_compressible.h:56]: Done.

File: tests/multicomponent_averaging/solution/solution-00000.0000.gnuplot
Comment (bobmyhill) [reply tests/multicomponent_averaging/solution/solution-00000.0000.gnuplot:8]: Thank you for spotting this! This is a significant bug.

The equation of state has been broken at some time after I made it. The reference temperature for the equation of state has been made a function of adiabatic heating. This is wrong; the reference temperature is a fixed constant in this equation of state. 

The adiabatic gradient should be determined by the equation of state, not the other way round. 
For now, I've returned the EoS to its intended form, but do we need to discuss the interaction with the adiabatic heating plugin?

File: source/material_model/multicomponent.cc
Comment (bobmyhill) [reply source/material_model/multicomponent.cc:186]: Thanks! I think it's still worth specifying how the compressibilities would be averaged if they were not zero, but I've added a few extra words to make it clear that this model is incompressible.

File: tests/multicomponent_compressible_averaging/solution/solution-00000.0000.gnuplot
Comment (bobmyhill) [reply tests/multicomponent_compressible_averaging/solution/solution-00000.0000.gnuplot:9]: The text at the top of the prm only listed the expected isochoric specific heat, which is not returned by ASPECT. I've updated the text to include the predicted isobaric specific heat, which is 1527.3 J/K/kg, as output by the test.

**Review by jdannberg — COMMENTED** [review]
This looks good to me now except for the problem with the incompressible model in the case adiabatic heating is used.

File: include/aspect/material_model/utilities.h
Comment (gassmoeller) [inline include/aspect/material_model/utilities.h:244]: I have a question and a comment here:
1. Would it make sense to name this `compute_composition_fractions`? This would be closer to the original name, and short, but carry the same meaning as far as I understand.
2. Please leave a copy of the old function for now, and deprecate it. The function implementation would then just call the new function. This is just because we usually do not immediately remove a function, but wait for at last one release before removing/renaming them. The old function would then look like this:
```
/**
* See compute_composition_fractions() for the documentation of this function.
* @deprecated: This function is deprecated. Please use compute_composition_fractions() instead.
       */
DEAL_II_DEPRECATED
      std::vector<double>
      compute_volume_fractions(const std::vector<double> &compositional_fields,
                               const ComponentMask &field_mask = ComponentMask());
```

File: source/material_model/equation_of_state/multicomponent_incompressible.cc
Comment (gassmoeller) [inline source/material_model/equation_of_state/multicomponent_incompressible.cc:48]: This needs to stay. See #3896.

File: source/material_model/utilities.cc
Comment (gassmoeller) [inline source/material_model/utilities.cc:728]: Not very important, but maybe like this? I do not care very much, but field fractions may be confusing if the compositions were mapped from particles (users could wonder why this is `field`).

```suggestion
        std::vector<double> composition_fractions(compositional_fields.size()+1);
```

File: tests/multicomponent_compressible_averaging.prm
Comment (gassmoeller) [inline tests/multicomponent_compressible_averaging.prm:6]:
```suggestion
# which equates to a volume fraction of 0.6 and 0.4.
```

**Review by gassmoeller — COMMENTED** [review]
Thanks for the fixes. I generally agree with the changes, I only have a few questions and smaller comments.

**Comment by bobmyhill** [comment]
Thanks for the suggestions Rene! I've made all the changes you requested. The issue with the multicomponent incompressible model is now dealt with in #3961.

File: tests/multicomponent_averaging/solution/solution-00000.0000.gnuplot
Comment (bobmyhill) [reply tests/multicomponent_averaging/solution/solution-00000.0000.gnuplot:8]: Thank you for spotting this! This is due to changes in multicomponent incompressible, which will be changed further in #3961. There's nothing wrong with this output; the behaviour is unexpected because the new functionality is not yet documented. Again, this is rectified in #3961.

**Comment by naliboff** [comment]
I agree with these changes and thanks for taking this on @bobmyhill. It looks like this has already been thoroughly reviewed, but I made a number of comments in #3961.

File: tests/multicomponent_averaging/solution/solution-00000.0000.gnuplot
Comment (jdannberg) [reply tests/multicomponent_averaging/solution/solution-00000.0000.gnuplot:8]: But the test output should reflect what your comment in the prm file says. You say that at the reference temperature of 273 , the density should be 2400, but it's not. Wouldn't it make more sense to make this a test without adiabatic heating (so the problem doesn't occur), and then make a separate test for the new functionality in #3961?

**Review by jdannberg — COMMENTED** [review]
Except for the one test I commented on, this looks good to me now.

**Comment by bobmyhill** [comment]
Rats! You're right, and I've now modified the test such that #3961 shouldn't break it <touch wood>. Thanks!

**Review by jdannberg — APPROVED** [review]
Alright, looks good to me. To clarify, #3961 will change the test results (since the test includes adiabatic heating) but will not change the first line that has the results described in the prm. 

That still leaves my question for the other maintainers if we all think that mass fraction averaging is the right thing to do in incompressible models. For compressible models, this is clearly a bugfix, for incompressible models, I think it's a matter of opinion what averaging should be used, and this change makes it more consistent for one given material model, but changes the results. Do we all agree this is what we want? @gassmoeller?

Other than that, this is okay to merge from my side. Thank you again for your work on this!

**Comment by jdannberg** [comment]
I think we have to think about this a bit more. I was just talking about this with @gassmoeller, and I think the Boussinesq Approximation internally assumes that compositional fields are volume fractions (the buoyancy term is Delta rho * C, which implies C is a volume fraction). So it's worth it to think about this a bit more before we decide we want to change all incompressible models (at least we would have to make sure that all benchmarks still work). Maybe it's a useful point to discuss at the next user meeting.

**Comment by bobmyhill** [comment]
> To clarify, #3961 will change the test results (since the test includes adiabatic heating) but will not change the first line that has the results described in the prm.
#3961 should only change tests without adiabatic heating.

> That still leaves my question for the other maintainers if we all think that mass fraction averaging is the right thing to do in incompressible models.
I'm ok either way. Averaging in this way does come with an additional computational cost, which users might want to avoid (especially if they mostly solve problems where compositional fractions are either 1 or 0). And it's not like we can ever claim "consistency" in an incompressible model.

**Comment by bobmyhill** [comment]
> the buoyancy term is Delta rho * C, which implies C is a volume fraction
If that is correct, it would be better to remove the changes from multicomponent incompressible. Good catch @gassmoeller! 
There's also the beta_T and beta_S issue. Maybe these are good points to add to the manual?

**Review by gassmoeller — APPROVED** [review]
Awesome, thanks! Actually @jdannberg figured out the inconsistency with the Boussinesq approximation and just asked for my opinion afterwards. But either way, I agree that the PR represents a clear improvement over the current state.
This looks good to go to me (@jdannberg: want to do the honor of merging? You did all of the review work).

**Comment by bobmyhill** [comment]
Hurrah! Thanks @gassmoeller, thanks @jdannberg!

**Comment by gassmoeller** [comment]
Thanks, this looks good, but the `compute_volume_fractions` function is missing an explicit instantiation at the end of the file. That is why the tests fail.

**Comment by bobmyhill** [comment]
Thanks @gassmoeller, now fixed.

**Review by gassmoeller — APPROVED** [review]
Looks good. Thanks again!

**Pull request merged**
